base/reducedim.jl

```diff
@@ -144,10 +144,18 @@ for (f1, f2, initval) in ((:min, :max, :Inf), (:max, :min, :(-Inf)))
             # otherwise use the min/max of the first slice as initial value
             v0 = mapreduce(f, $f2, A1)
 
-            # but NaNs need to be avoided as initial values
-            v0 = v0 != v0 ? typeof(v0)($initval) : v0
-
             T = _realtype(f, promote_union(eltype(A)))
+
+            # but NaNs and missing need to be avoided as initial values
+            if (v0 != v0) === true
+                v0 = typeof(v0)($initval)
+            elseif ismissing(v0)
+                # If it's a union type, pick the initval from the other type.
+                if typeof(T) == Union
+                    v0 = (T.a == Missing ? T.b : T.a)($initval)
+                end
+            end
```

Comment:
I think `Missing` is always `T.a` if it's in a Union, it's probably `type_morespecific` than anything but `Any`, but I didn't want to rely on that.

I was also too tired to try to work out if I could avoid defining a `v0` altogether, though I'm not hopeful about that.

Reply:
Missing is not necessary always T.a AFAIK it puts singleton to the first position and Missing is often the only singleton but this is not guaranteed (there's no intended behavior in this)

Reply:
You could also do `nonmissingtype(T)($initval)`. Actually I'm not even sure you need to check that the `T` is a `Union` nor to extract the type, as `convert(T, $initval)` should work thanks to special `convert` methods defined for `Missing`.

BTW, this mostly replaces https://github.com/JuliaLang/julia/pull/35323 (which is less efficient but also works when the type isn't `Union{T, Missing}`).

Reply:
Using `convert` seems like a good idea. Perhaps this?

```julia
            # but NaNs and missing need to be avoided as initial values
            if is_poisoning(v0)
                # convert handles Missing specially.
                v0 = convert(T, $initval)
            end
```

Reply:
The previous code what as a bit more robust because it only special cases `NaN` and then inserts `Inf` or `-Inf`, but for unions with missing `T` could be anything. I wonder if this is better:

```jl
            if is_poisoning(v0)
                v0 = T == Any ? $initval : typemin(nonmissingtype(T))
            end
```

Or, if `nonmissingtype` isn't optimised out by the compiler:

```jl
            if ismissing(v0)
                v0 = T == Any ? $initval : typemin(nonmissingtype(T))
            elseif (v0 == v0) !== true
                # v0 is NaN
                v0 = $initval
            end
```

Reply:
Actually I'm not sure either solution is correct. `initval` (which is `Inf` or `-Inf`) may not be convertible to `T` (e.g. for integers), and `typemin`/`typemax` may be undefined (e.g. for `BigInt` or abstract types). That's why at https://github.com/JuliaLang/julia/pull/35323 I adopted an inefficient but robust approach. Maybe we could use `$initval` when `nonmissingtype(T) <: AbstractFloat` just like the `NaN` path does, use `typemin`/`typemax` if it doesn't throw an error, and otherwise use the approach from https://github.com/JuliaLang/julia/pull/35323.

Reply:
Good catch. What a pain. Just introducing a `try/catch` will slow things down, though maybe the typical size of `A` is large enough that it doesn't matter.

Reply:
Maybe this code could be re-written so that it doesn't need an initial value? Or it just uses the first value or something?

Reply:
Maybe something like this:

```jl
            if is_poisoning(v0)
                Tnm = nonmissingtype(T)
                if Tnm <: AbstractFloat
                    v0 = $initval
                else
                    try
                        v0 = $typeextreme(Tnm)
                    catch e
                        e isa MethodError || rethrow()
                        v0 = mapreduce(f, $f2, skipmissing(A))
                    end
                end
            end
```

I'd be happy to just leave `minimum(A; dims)` broken on `missing` and just get this PR merged.

Reply:
Looks like it. But indeed you could leave that out of this PR, anyway it's been broken for a long time.

Reply:
That's still not right because mapreduce will error if `skipmissing(A)` is empty.

I think this is right, but it's not pretty.

```jl
            T = _realtype(f, promote_union(eltype(A)))
            Tr = v0 isa T ? T : typeof(v0)

            # but NaNs and missing need to be avoided as initial values
            if is_poisoning(v0)
                Tnm = nonmissingtype(Tr)
                if Tnm <: AbstractFloat
                    v0 = $initval
                else
                    try
                        v0 = $typeextreme(Tnm)
                    catch e
                        e isa MethodError || rethrow()
                        # Find the min/max value in the array,
                        # skipping missing.
                        for v in A
                            fv = f(v)
                            if !ismissing(fv)
                                v0 = ismissing(v0) ? fv : $f2(v0, fv)
                            end
                        end
                    end
                end
            end
```